Implement form to search and add users to groups

T-Dnzt opened the pull request:
# Status

Ready for review.

# Related Issue

 [#72 Add the "Add Member" feature](https://github.com/greencommons/commons/issues/72)

# Summary

This pull request adds the search form to add users to groups. It comes with an autocomplete feature that can be easily reused by other views in the future. The autocomplete is only based on a `ILIKE` PG query using the `email` column of the `users` table. We will need to add the user names later as well. If you think performances will be an issue, I could use Elastic Cache to keep an index instead of just querying PG (this can be done in a future PR I believe).

# Changes

- Addition of an Ajax loader gif to show the users that suggestions are loading.
- Addition of the [JQuery Autocomplete](http://easyautocomplete.com/) library. 
- Addition of the JQuery Autocomplete css. I didn't put it in the `vendor/` folder because I made some changes there (removed what we didn't need). I also decided to put in under `themes/` since I feel it should be part of the base theme of the app.
- Creation of the `AutocompleteController` with a `users` action.
- Creation of a `UserFilter` service that will filter the users based on the given query and remove all users that are already members of the specified group. This service can be easily be expanded and take more arguments in the future if needed.
- Addition of the `create` action on the MembersController to receive the email submitted by the autocomplete form.
- Addition of a some RSpec helper to test `JSON` endpoints like the autocomplete one.

# Screenshots

Searching for email:
<img width="609" alt="screen shot 2559-12-13 at 11 37 58 pm" src="https://cloud.githubusercontent.com/assets/2647689/21149329/c1b64844-c18d-11e6-84f2-e52ffa114fc2.png">

Picking an email (the email can also just be pasted directly and the autocomplete will only validate that this email exists):
<img width="607" alt="screen shot 2559-12-13 at 11 38 21 pm" src="https://cloud.githubusercontent.com/assets/2647689/21149332/c315ac02-c18d-11e6-84bb-bb944a948280.png">

After clicking on the + button:
<img width="590" alt="screen shot 2559-12-13 at 11 38 25 pm" src="https://cloud.githubusercontent.com/assets/2647689/21149339/c5fbca1e-c18d-11e6-85dd-3a88a1d05492.png">

If no user was found:
<img width="585" alt="screen shot 2559-12-13 at 11 45 51 pm" src="https://cloud.githubusercontent.com/assets/2647689/21149531/54dbf7a4-c18e-11e6-9f7d-84ace502dc68.png">

# Style

I noticed we were using Rubocop and I've started fixing the code based on its suggestions. I decided to leave them out of this pull request and I will create a new one later that will fix all the offenses in the Group Feature branches.

adarsh commented:
> I noticed we were using Rubocop and I've started fixing the code based on its suggestions. I decided to leave them out of this pull request and I will create a new one later that will fix all the offenses in the Group Feature branches.

If you're interested in catching these before they make it into a PR, run `bundle exec rake` instead of `rake spec` to also run Rubocop against your branch. 🚓

ptrikutam commented on app/assets/javascripts/autocomplete.js:
```diff
@@ -0,0 +1,37 @@
+$(document).ready(function() {
```
Would be great to either convert this to React or Vue.js code for now -- I can see this class simplifying a lot and living alongside the markup (potentially CSS code too). 

That can totally come at a later time.

T-Dnzt replied:
I think changing it now is best, just so we can set the right example. I'll switch to React :)

ptrikutam commented on app/assets/stylesheets/application.scss:
```diff
@@ -2,6 +2,7 @@
 @import "theme/variables";
 @import "bootstrap";
 @import "theme/bootswatch";
+@import "theme/autocomplete";
```
Would it actually be better to keep this to the `/components` folder? I would imagine we should keep code in `/theme` that we can easily overwrite with a vendor theme of sorts (something from [Bootswatch](https://bootswatch.com/)), and since we have custom code that wouldn't be doable here.

Also, if we have the CSS /JS / Markup all with the name of `autocomplete` it'll be easier to identify as a cohesive component.

What do you think?

ptrikutam commented on app/controllers/api/autocomplete_controller.rb:
```diff
@@ -0,0 +1,11 @@
+module Api
```
@dweinberger is going to love this :)

@dweinberger -- this is the first class in the API for Green Commons -- we can start building out additional API endpoints from here on out, but this is our first!

ptrikutam commented on app/controllers/api/autocomplete_controller.rb:
```diff
@@ -0,0 +1,11 @@
+module Api
+  class AutocompleteController < ApplicationController
+    before_action :authenticate_user!
```
This is totally fine for for now -- down the line we may want to introduce some kind of OAuth for the API. Might add complexity for simple components that just need to query the back end really quick, though...

T-Dnzt replied:
Totally. For now, it's good enough like this, but we will probably have to change it later on. I don't know much details about the web API that's needed in the project (public-facing, only for web app or available for other private clients, etc). We will need to talk about this.

ptrikutam commented on app/controllers/members_controller.rb:
```diff
@@ -1,14 +1,20 @@
-  before_action :set_group_user, except: [:index]
+  before_action :set_group_user, except: [:index, :create]
   before_action :set_is_admin
   before_action :check_admin_rights, except: [:index]
```
Do we need an `:except` for `:create` here too?

T-Dnzt replied:
Hmmm, only group admins can add new users, right? If that's the case, we need to run the `check_admin_rights` before allowing the addition of a new member.

ptrikutam replied:
Ah, my mistake -- for some reason I thought this was the `GroupsController` when I wrote that comment. Please ignore :)

ptrikutam commented on app/models/group.rb:
```diff
@@ -13,11 +13,11 @@ def resources
 
   def add_admin(user)
-    groups_users.create(user: user, admin: true)
+    groups_users.new(user: user, admin: true).save
```
Curious about the motivation for this change?

T-Dnzt replied:
Alright, I'm going to be completely honest - I don't remember. I had a reason, maybe even a good one, but I'm not sure why this change was needed. I'm going to look into it and remove it if it's not needed. Else, I'll let you know why I changed it. Deal? ;)

T-Dnzt replied:
Yes, please see my answer to Pavan. I will get back to you about this change ;)

T-Dnzt replied:
Okay, I remember! It was because of this:

```ruby
if @group.add_user(user)
  'Member was successfully added.'
else
  'Member already in group.'
end
```

`create` always return the object, even if it couldn't be saved. Using `.save` either returns `true` or `false` which I can use to show the appropriate message.

ptrikutam replied:
Makes sense. I'm good with this code.

ptrikutam commented on app/service/user_filter.rb:
```diff
@@ -0,0 +1,21 @@
+
+  def group_members
+    return [] unless @group_id
+    @group_members ||= User.joins(:groups_users).where('group_id = ?', @group_id)
```
Could we not do something like `Group.find(@group_id).members`?

ptrikutam replied:
More of a question about the best way to run this query, it's a nitpicky comment and not a huge deal.

T-Dnzt replied:
I actually started with something like this. I even had a `group` method that was loading the group. But the request was then running 3 queries (get users, get group, get users from group). I thought it was better to optimize it a bit and reduce to the only two queries we really need: getting users and getting the users belonging to a `group_id`. 

In the case of a non-existing group, the performances would be better with the third request to get the group, because it would mean we don't need to run the get members one. However, if the group exists (which should be the case most of the time when using this), we're saving a tiny bit of time by not querying the group.

I hope it makes sense. I don't mind changing it to query the group if you think it's better.

ptrikutam replied:
OK, good enough for me. No need for a change here :)

ptrikutam commented on spec/requests/api/autocomplete_controller_spec.rb:
```diff
@@ -0,0 +1,70 @@
+          it 'returns only the users that do not belong in the group' do
+            group.add_user(john)
+            group.add_user(mark)
+            jack
```
Instead of listing the variables here, you can actually use `let!(:jack)` at the top so that it creates the user automatically, even before it's accessed later in the spec. Unless of course you prefer them not to be created until they're needed within a spec, in which case you can ignore this comment.

adarsh replied:
Personal Preference: I often get confused with the style of testing where `let`s are used to DRY up test code. In this case, a declaration like `jack` is difficult to parse on first blush.

I'd almost prefer to just call `jack = create(:user, email: 'jack@commons.org')` once in each test, rather than have Mystery Guests.

More on why [here][1].

[1]: https://robots.thoughtbot.com/lets-not

T-Dnzt replied:
Yes, I actually prefer to only create them if they are needed. However, I think @adarsh has a good point and I'm going to change these tests to avoid mystery guests.

T-Dnzt replied:
I've read this article before and I promised myself I would stop doing this! But the habits won, and I did it again. I agree with you and also prefer to see tests that are easy to understand at a quick glance. Therefore, I will follow your recommendations and simply create them in the tests.

adarsh commented on spec/service/user_filter_spec.rb:
```diff
@@ -0,0 +1,70 @@
+
+  describe '#run' do
+    context 'when group_id is given to scope down the results' do
+      context 'when query = comm' do
```
I'm not sure we need this context block - it feels clear in the test.

adarsh commented on spec/service/user_filter_spec.rb:
```diff
@@ -0,0 +1,70 @@
+        end
+      end
+
+      context 'when query = jo' do
```
Same here, this is clear in the test code.

adarsh commented on spec/support/helpers.rb:
```diff
@@ -0,0 +1,9 @@
+module Helpers
```
Can we rename this to something like `JSONHelpers`? I've seen generically named "helper" classes bloat to 800+ lines ⛱

T-Dnzt replied:
Haha, yes, I'm going to do that. Nice catch!

ptrikutam commented on app/assets/stylesheets/theme/_autocomplete.scss:
```diff
@@ -0,0 +1,101 @@
+  position: relative;
+}
+.easy-autocomplete input {
+  border-color: #ccc;
```
Good call @adarsh. Actually, taking it a step further, we should either use colors defined in the [`_variables.scss`](https://github.com/greencommons/commons/blob/master/app/assets/stylesheets/theme/_variables.scss), or add something to it. For this, maybe one of the `$gray-*` colors?

T-Dnzt commented on app/assets/javascripts/autocomplete.js:
```diff
@@ -0,0 +1,37 @@
+    var path = input.data("autocomplete");
+
+    input.on('input', function() {
+      if(input.val().length > 0) {
```
Agreed, but I couldn't figure out a way to do that with this JQuery plugin. The compromise I found was the `requestDelay` option which will wait for 500ms before sending the request, letting the time to a user to type a certain amount of characters.

This might not be relevant anymore once I switch to React and if I can, I will set a minimum amount of characters before starting the autocomplete.

T-Dnzt commented on app/assets/javascripts/autocomplete.js:
```diff
@@ -0,0 +1,37 @@
+      theme: "bootstrap"
+    });
+  });
+
```
Hmm, nope. Will fix ;)

T-Dnzt commented on app/controllers/members_controller.rb:
```diff
@@ -49,4 +55,10 @@ def check_admin_rights
     end
+
+    def add_group_user(user)
+      return 'User not found.' unless user
```
Used to be like this, and I changed it. Oops. I think I will put it back if you think it's less confusing. I kinda find it easier to read the flow than with nested `if...else` but I get your point.

T-Dnzt commented on app/controllers/api/autocomplete_controller.rb:
```diff
@@ -0,0 +1,11 @@
+      users = UserFilter.new(query: params[:q], group_id: params[:group_id]).run
+      render json: users.map { |u| { email: u.email } }
+    end
+
```
Thanks, will fix.

T-Dnzt commented:
Alright guys, so I think I implemented all the requested changes. I also switched the whole thing form JQuery to React... Now I'm a bit rusty with React but I did my best to create a logical component named `AutocompleteForm`. I also extracted some of the logic to `AutocompleteSuggestions` that deals with the AJAX call and showing the list of results. 

I'm totally open to discussion, changes and suggestions about the best ways to implement this component and don't mind fixing it if needed. 

Implementing this also makes me think we could spice up the members management page with some React component, making it more dynamic overall. If we decide to go that way, it will be done in a future PR but I will create a new issue for it.

/discuss

T-Dnzt commented on app/views/members/index.html.slim:
```diff
@@ -3,21 +3,20 @@
     
 .row
-  .col-lg-6
+  .col-xs-12.col-md-6.col-lg-6
```
Yes I know... My mistake for doing it desktop-first and forgetting to remove le lg class :) I will fix this before merging.

T-Dnzt commented:
I will be waiting for [PR #84](https://github.com/greencommons/commons/pull/84) to be merged before merging this one. Once it has been merged, I will be able to fix the remaining Rubocop issues only present in the changes. @adarsh I also need to know if you're planning on doing another review or if I should dismiss your review? cc: @ptrikutam

adarsh commented:
Don't wait on me for anything. Merge away at your discretion.


On Thu, Dec 15, 2016 at 8:57 PM Thibault <notifications@github.com> wrote:

> I will be waiting for PR #84
> <https://github.com/greencommons/commons/pull/84> to be merged before
> merging this one. Once it has been merged, I will be able to fix the
> remaining Rubocop issues only present in the changes. @adarsh
> <https://github.com/adarsh> I also need to know if you're planning on
> doing another review or if I should dismiss your review? cc: @ptrikutam
> <https://github.com/ptrikutam>
>
>
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greencommons/commons/pull/73#issuecomment-267517098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3xXX-Wy4RJMSqOQGNvlH2LwiclX8Zlks5rIhpVgaJpZM4LKALg>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
> --
Sent from my mobile
